cpp/src/search/contains_table.cu

```diff
@@ -1,148 +1,408 @@
-  return contained;
-}
-
-}  // namespace cudf::detail
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <join/join_common_utils.cuh>
+
+#include <cudf/detail/join.hpp>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/structs/utilities.hpp>
+#include <cudf/table/experimental/row_operators.cuh>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <thrust/iterator/counting_iterator.h>
+
+#include <cuco/static_map.cuh>
+
+#include <type_traits>
+
+namespace cudf::detail {
+
+namespace {
+
+using cudf::experimental::row::lhs_index_type;
+using cudf::experimental::row::rhs_index_type;
+
+using static_map = cuco::static_map<lhs_index_type,
+                                    size_type,
+                                    cuda::thread_scope_device,
+                                    rmm::mr::stream_allocator_adaptor<default_allocator<char>>>;
+
+/**
+ * @brief Check if the given type `T` is a strong index type (i.e., `lhs_index_type` or
+ * `rhs_index_type`).
+ *
+ * @return A boolean value indicating if `T` is a strong index type
+ */
+template <typename T>
+constexpr auto is_strong_index_type()
+{
+  return std::is_same_v<T, lhs_index_type> || std::is_same_v<T, rhs_index_type>;
+}
+
+/**
+ * @brief An adapter functor to support strong index types for row hasher that must be operating on
+ * `cudf::size_type`.
+ */
+template <typename Hasher>
+struct strong_index_hasher_adapter {
+  strong_index_hasher_adapter(Hasher const& hasher) : _hasher{hasher} {}
+
+  template <typename T, CUDF_ENABLE_IF(is_strong_index_type<T>())>
+  __device__ constexpr auto operator()(T const idx) const noexcept
+  {
+    return _hasher(static_cast<size_type>(idx));
+  }
+
+ private:
+  Hasher const _hasher;
+};
+
+/**
+ * @brief An adapter functor to support strong index type for table row comparator that must be
+ * operating on `cudf::size_type`.
+ */
+template <typename Comparator>
+struct strong_index_comparator_adapter {
+  strong_index_comparator_adapter(Comparator const& comparator) : _comparator{comparator} {}
+
+  template <typename T,
+            typename U,
+            CUDF_ENABLE_IF(is_strong_index_type<T>() && is_strong_index_type<U>())>
+  __device__ constexpr auto operator()(T const lhs_index, U const rhs_index) const noexcept

```

Comment:
@ttnghia Can you explain what types this functor expects to get? `lhs_index_type, rhs_index_type` and `rhs_index_type` or also `size_type, size_type`? I want to simplify this but I don't fully understand the requirements you're aiming for here because the templates are fairly broad.

Reply:
`(size_type, size_type)` will never be instantiated, because only strong index types are used in the code. The template function will not be enabled for this case.

Only `(lhs_index_type, lhs_index_type)` and `(lhs_index_type, rhs_index_type)` will be instantiated in this file, at this time:
 * `(lhs_index_type, lhs_index_type)` will be used for self-comparing the haystack table, when inserting row indices of that table into the map.
 * `(lhs_index_type, rhs_index_type)` will be used for checking contains using `cuco::static_map::contains`. The internal implementation of `cuco::static_map::contains` at this moment only call `(lhs_index_type, rhs_index_type)`, but it doesn't guarantee to not call `(rhs_index_type, lhs_index_type)` in the future. Nevertheless, this function will flip the order to make sure the right indices will be passed into the underlying comparator.

Reply:
Could we use three explicit `operator()`s, one that addresses the self-comparison `lhs_index_type, lhs_index_type` case, and two that address the left/right (right/left) pairings? This template logic is quite complicated. I propose:

```cpp
/**
 * @brief An adapter functor to support strong index type for table row comparator that must be
 * operating on `cudf::size_type`.
 */
template <typename Comparator>
struct strong_index_comparator_adapter {
  strong_index_comparator_adapter(Comparator const& comparator) : _comparator{comparator} {}

  __device__ constexpr auto operator()(lhs_index_type const left, lhs_index_type const right) const noexcept
  {
    return _comparator(static_cast<size_type>(left), static_cast<size_type>(right);
  }

  __device__ constexpr auto operator()(lhs_index_type const left, rhs_index_type const right) const noexcept
  {
    return _comparator(static_cast<size_type>(left), static_cast<size_type>(right);
  }

  __device__ constexpr auto operator()(rhs_index_type const right, lhs_index_type const left) const noexcept
  {
    // Reverse the order to left, right.
    return this->operator()(left, right);
  }

 private:
  Comparator const _comparator;
};
```

Reply:
Sounds good.

Reply:
Wait, no. I have tried that. This will cause compiler warning: defined but never refered.

Reply:
There will be warnings because this struct will be instantiated with different underlying row comparator. Depending on situations, only one overload of `operator()` will be instantiated.

Reply:
I also tried `[[maybe_unused]]` with these overloads but it seems that `[[maybe_unused]]` is only applied to the returning value, not to the function body.

Reply:
Can you post the full warning output?

Reply:
```
../src/search/contains_table.cu(92): warning #177-D: function "cudf::detail::<unnamed>::strong_index_comparator_adapter<Comparator>::operator()(cudf::experimental::row::lhs_index_type, cudf::experimental::row::rhs_index_type) const noexcept [with Comparator=cudf::experimental::row::equality::device_row_comparator<cudf::nullate::DYNAMIC, cudf::experimental::row::equality::physical_equality_comparator>]" was declared but never referenced

../src/search/contains_table.cu(98): warning #177-D: function "cudf::detail::<unnamed>::strong_index_comparator_adapter<Comparator>::operator()(cudf::experimental::row::rhs_index_type, cudf::experimental::row::lhs_index_type) const noexcept [with Comparator=cudf::experimental::row::equality::device_row_comparator<cudf::nullate::DYNAMIC, cudf::experimental::row::equality::physical_equality_comparator>]" was declared but never referenced

../src/search/contains_table.cu(92): warning #177-D: function "cudf::detail::<unnamed>::strong_index_comparator_adapter<Comparator>::operator()(cudf::experimental::row::lhs_index_type, cudf::experimental::row::rhs_index_type) const noexcept [with Comparator=cudf::experimental::row::equality::device_row_comparator<cudf::nullate::DYNAMIC, cudf::experimental::row::equality::nan_equal_physical_equality_comparator>]" was declared but never referenced

../src/search/contains_table.cu(98): warning #177-D: function "cudf::detail::<unnamed>::strong_index_comparator_adapter<Comparator>::operator()(cudf::experimental::row::rhs_index_type, cudf::experimental::row::lhs_index_type) const noexcept [with Comparator=cudf::experimental::row::equality::device_row_comparator<cudf::nullate::DYNAMIC, cudf::experimental::row::equality::nan_equal_physical_equality_comparator>]" was declared but never referenced
```

Reply:
🤷‍♂️ I don't want to push any more on this right now, but something seems off here.

Reply:
@ttnghia @bdice I think the reason that we don't need either of the extra overloads is the fact that the adapter is specifically designed to work with the `self_comparator` for the purpose of handling strong typing. @ttnghia listed two cases above for the comparators: 1) inserting into the map, and 2) checking whether a key is contained in the map. In the second case, however, we use the [`two_table_comparator`](https://github.com/rapidsai/cudf/pull/11330/files#diff-da6507d289bfdfeb6232555cd9ead614ad047a0ae9525b379cc89da3a7bb2ce0R248), which [already handles the strong type indices correctly](https://github.com/rapidsai/cudf/blob/branch-22.08/cpp/include/cudf/table/experimental/row_operators.cuh#L1313). We should _only_ need to handle the `(lhs_index_type, lhs_index_type)` case using the adapter AFAICT.

Reply:
That sounds like a good explanation to me! I couldn’t quite piece that together myself but I think it sounds accurate.

Reply:
Sorry that is just half correct. Here we have two separate code paths: 1) A code path uses exprimental row operators (`two_table_comparator`) and 2) Another code path uses the traditional `cudf::row_equality` for comparing rows from two tables ([row_equality](https://github.com/rapidsai/cudf/pull/11330/files#diff-da6507d289bfdfeb6232555cd9ead614ad047a0ae9525b379cc89da3a7bb2ce0R358)). This `cudf::row_equality` doesn't support strong index types thus it must be wrapped in this adapter as in the link.

Reply:
The reason why there are warnings is that: this adapter is a template struct, which means the compiler creates a separate version of when it is used. In this file, it is used 3 times thus there will be 3 different versions of this template struct will be created. For each version, only one overload of `operator()` is called (out of 3), causing warnings.

Reply:
Okay, that is helpful to know. Perhaps that indicates that we should split this struct into multiple structs which only have the functions used in each case.

Reply:
Ah good catch @ttnghia, thanks for the correction.